File: C#/src/Implementations/InputValidatorManager.cs
using System;
using System.Text.RegularExpressions;

public static class InputValidatorManager
{
    public static bool IsValidFileName(string fileName)
    {
        Regex pattern = new Regex("^[a-zA-Z0-9_-]{1,30}$");
        return pattern.IsMatch(fileName);
    }

    public static bool IsValidMenuChoice(string input)
    {
        Regex pattern = new Regex("^[0-6]$");
        return pattern.IsMatch(input);
    }

    public static bool IsValidItemChoice(string input)
    {
        Regex pattern = new Regex("^[0-6]$");
        return pattern.IsMatch(input);
    }

    public static bool IsValidYesNo(string input)
    {
        Regex pattern = new Regex("^[tTnN]$");
        return pattern.IsMatch(input);
    }

    public static bool IsValidDifficultyChoice(string input)
    {
        Regex pattern = new Regex("^[0-2]$");
        return pattern.IsMatch(input);
    }

    public static string GetValidatedInput(string prompt, Func<string, bool> validator)
    {
        string input;
        while (true)
        {
            Console.Write(prompt);
            input = Console.ReadLine() ?? "";

            if (validator(input))
            {
                return input;
            }
            Console.WriteLine("Nieprawidlowe dane! Sprobuj ponownie.");
        }
    }
}

File: C++/src/Implementations/InputValidatorManager.cpp
#include "InputValidatorManager.h"
#include "GameEnums.h"
#include <iostream>
#include <string>
#include <regex>
#include <thread>
#include <conio.h>

bool InputValidatorManager::IsValidFileName(const std::string& fileName)
{
    std::regex pattern("^[a-zA-Z0-9_-]{1,30}$");
    return regex_match(fileName, pattern);
}

bool InputValidatorManager::IsValidMenuChoice(const std::string& input)
{
    std::regex pattern("^[0-6]$");
    return regex_match(input, pattern);
}

bool InputValidatorManager::IsValidItemChoice(const std::string& input)
{
    std::regex pattern("^[0-6]$");
    return regex_match(input, pattern);
}

bool InputValidatorManager::IsValidYesNo(const std::string& input)
{
    std::regex pattern("^[0-1]$");
    return regex_match(input, pattern);
}

bool InputValidatorManager::IsValidDifficultyChoice(const std::string& input)
{
    std::regex pattern("^[0-2]$");
    return regex_match(input, pattern);
}

std::string InputValidatorManager::GetValidatedInput(const std::string& prompt,
    bool (*validator)(const std::string&))
{
    while (_kbhit()) _getch();

    std::cout << prompt << std::flush;
    std::string input = "";
    while (true) 
    {
        if (_kbhit()) 
        {
            char ch = _getch();
            if (ch == 13) 
            {
                std::cout << '\n';
                if (validator(input))
                {
                    while (_kbhit()) _getch();
                    return input;
                }
                else
                {
                    input = "";
                    std::cout << "Nieprawidlowy wybor. Sprobuj ponownie: ";
                }
            }
            else if (ch == 8 && !input.empty()) 
            {
                input.pop_back();
                std::cout << "\b \b";
            }
            else if (ch >= 32 && ch < 127) 
            {
                input += ch;
                std::cout << ch;
            }
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(16));
    }
}